File: FrontVantardSoftwareStudio/src/kernel/crypto.js
const NONCE_SIZE = 12;

let cachedKeyPromise = null;

function bytesToBase64(bytes) {
  let binary = "";
  const chunkSize = 0x8000;

  for (let i = 0; i < bytes.length; i += chunkSize) {
    binary += String.fromCodePoint(...bytes.subarray(i, i + chunkSize));
  }

  return globalThis.btoa(binary);
}

function base64ToBytes(b64) {
  const binary = globalThis.atob(b64);
  const bytes = new Uint8Array(binary.length);

  for (let i = 0; i < binary.length; i++) {
    const codePoint = binary.codePointAt(i);
    bytes[i] = codePoint ?? 0;
  }

  return bytes;
}

function concatBytes(a, b) {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

async function getAesGcmKey() {
  if (!globalThis.crypto?.subtle) {
    throw new Error("WebCrypto no está disponible en este entorno.");
  }

  const rawKeyStr = import.meta.env.VITE_AES_KEY;
  if (!rawKeyStr) {
    throw new Error("Falta configurar VITE_AES_KEY.");
  }

  if (!cachedKeyPromise) {
    const rawKeyBytes = new TextEncoder().encode(rawKeyStr);
    cachedKeyPromise = globalThis.crypto.subtle.importKey(
      "raw",
      rawKeyBytes,
      { name: "AES-GCM" },
      false,
      ["encrypt", "decrypt"],
    );
  }

  return cachedKeyPromise;
}

export async function encryptPayload(data) {
  const key = await getAesGcmKey();
  const iv = globalThis.crypto.getRandomValues(new Uint8Array(NONCE_SIZE));
  const plaintext = new TextEncoder().encode(JSON.stringify(data));

  const encrypted = await globalThis.crypto.subtle.encrypt(
    { name: "AES-GCM", iv },
    key,
    plaintext,
  );

  const combined = concatBytes(iv, new Uint8Array(encrypted));
  return bytesToBase64(combined);
}

export async function decryptPayload(ciphertextB64) {
  if (!ciphertextB64) {
    throw new Error("Ciphertext vacío.");
  }

  const key = await getAesGcmKey();
  const combined = base64ToBytes(ciphertextB64);
  if (combined.length < NONCE_SIZE + 16) {
    throw new Error("Ciphertext inválido o incompleto.");
  }

  const iv = combined.slice(0, NONCE_SIZE);
  const data = combined.slice(NONCE_SIZE);

  const decrypted = await globalThis.crypto.subtle.decrypt(
    { name: "AES-GCM", iv },
    key,
    data,
  );

  const plaintext = new TextDecoder().decode(decrypted);
  return JSON.parse(plaintext);
}

File: BackVantardSoftwareStudio/users/crypto.py
import base64
import json
import os

from Crypto.Cipher import AES
from django.conf import settings


NONCE_SIZE = 12
TAG_SIZE = 16


def _get_key():
    return settings.AES_SECRET_KEY.encode('utf-8')


def encrypt_payload(data: dict) -> str:
    key = _get_key()
    nonce = os.urandom(NONCE_SIZE)
    plaintext = json.dumps(data).encode('utf-8')
    cipher = AES.new(key, AES.MODE_GCM, nonce=nonce)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)
    return base64.b64encode(nonce + ciphertext + tag).decode('utf-8')


def decrypt_payload(ciphertext_b64: str) -> dict:
    key = _get_key()
    raw = base64.b64decode(ciphertext_b64, validate=True)
    if len(raw) < (NONCE_SIZE + TAG_SIZE):
        raise ValueError('Ciphertext inválido o incompleto.')

    nonce = raw[:NONCE_SIZE]
    tag = raw[-TAG_SIZE:]
    ciphertext = raw[NONCE_SIZE:-TAG_SIZE]

    cipher = AES.new(key, AES.MODE_GCM, nonce=nonce)
    plaintext = cipher.decrypt_and_verify(ciphertext, tag)
    return json.loads(plaintext.decode('utf-8'))
